src/EditorFeatures/CSharpTest/Completion/CompletionProviders/AwaitCompletionProviderTests.cs

```diff
@@ -279,5 +279,231 @@ public async Task TestAwaitInLock_TopLevel()
         }
+
+        [Fact]
+        public async Task TestDotAwaitSuggestAfterDotOnTask()
+        {
+            await VerifyKeywordAsync(@"
+using System.Threading.Tasks;
+
+class C
+{
+  async Task F(Task someTask)
+  {
+    someTask.$$
+  }
+}
+", LanguageVersion.CSharp9);
+        }
+
+        [Fact]
+        public async Task TestDotAwaitSuggestAfterDotOnValueTask()
+        {
+            var valueTaskAssembly = typeof(ValueTask).Assembly.Location;
+            await VerifyItemExistsAsync(@$"
+<Workspace>
+    <Project Language=""C#"" AssemblyName=""Assembly1"" CommonReferences=""true"" LanguageVersion=""{LanguageVersion.CSharp9.ToDisplayString()}"">
+        <MetadataReference>{valueTaskAssembly}</MetadataReference>
+        <Document FilePath=""Test2.cs"">
+using System.Threading.Tasks;
+
+class C
+{{
+  async Task F(ValueTask someTask)
+  {{
+    someTask.$$
+  }}
+}}
+        </Document>
+    </Project>
+</Workspace>
+", "await");
+        }
+
+        [Fact]
+        public async Task TestDotAwaitSuggestAfterDotOnCustomAwaitable()
+        {
+            await VerifyKeywordAsync(@"
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+
+public class DummyAwaiter: INotifyCompletion {
+    public bool IsCompleted => true;
+    public void OnCompleted(Action continuation) => continuation();
+    public void GetResult() {}
+}
+
+public class CustomAwaitable
+{
+    public DummyAwaiter GetAwaiter() => new DummyAwaiter();
+}
+
+static class Program
+{
+    static async Task Main()
+    {
+        var awaitable = new CustomAwaitable();
+        awaitable.$$;
+    }
+}", LanguageVersion.CSharp9);
+        }
+
+        [Fact]
+        public async Task TestDotAwaitSuggestAfterDotDot()
+        {
+            await VerifyKeywordAsync(@"
+using System.Threading.Tasks;
+
+static class Program
+{
+    static async Task Main(Task someTask)
+    {
+        someTask.$$.;
+    }
+}", LanguageVersion.CSharp9);
+        }
+
+        [Fact]
+        public async Task TestDotAwaitSuggestAfterDotBeforeType()
+        {
+            await VerifyKeywordAsync(@"
+using System;
+using System.Threading.Tasks;
+
+static class Program
+{
+    static async Task Main(Task someTask)
+    {
+        someTask.$$
+        Int32 i = 0;
+    }
+}", LanguageVersion.CSharp9);
+        }
+
+        [Fact]
+        public async Task TestDotAwaitSuggestAfterDotBeforeAnotherAwait()
+        {
+            await VerifyKeywordAsync(@"
+using System;
+using System.Threading.Tasks;
+
+static class Program
+{
+    static async Task Main(Task someTask)
+    {
+        someTask.$$
+        await Test();
+    }
+
+    async Task Test() { }
+}", LanguageVersion.CSharp9);
+        }
+
+        [Fact]
+        public async Task TestDotAwaitNotAfterDotOnTaskIfAlreadyAwaited()
+        {
+            await VerifyAbsenceAsync(@"
+using System.Threading.Tasks;
+
+class C
+{
+  async Task F(Task someTask)
+  {
+    await someTask.$$
```

Comment:
haven't looked through all tests, but can we have some more complex expression cases.  like `a.b.` and `foo().` and `(x + y).` etc.  basically, just ensrue that on lots of expr/type cases that things are working properly.

Reply:
Done in f2b6774e